Adjust selection to scroll dir

BrianHung opened the pull request:
Fixes #819 

Not 100% sure useEffect is the right way to adjust selection when scroll edge but seems to work.

jassmith commented on packages/core/src/internal/data-grid/data-grid.tsx:
```diff
@@ -510,10 +510,9 @@ const DataGrid: React.ForwardRefRenderFunction<DataGridRef, DataGridProps> = (p,
-                Math.abs(x) < edgeSize ? -1 : Math.abs(rect.width - x) < edgeSize ? 1 : 0,
-                Math.abs(y) < edgeSize ? -1 : Math.abs(rect.height - y) < edgeSize ? 1 : 0,
+                x < stickyX ? -1 : rect.width < x ? 1 : 0,
+                y < totalHeaderHeight ? -1 : rect.height < y ? 1 : 0,
```
You're missing deps for this callback.

jassmith commented on packages/core/src/data-editor/data-editor.tsx:
```diff
@@ -2091,6 +2091,59 @@ const DataEditorImpl: React.ForwardRefRenderFunction<DataEditorRef, DataEditorPr
     const [scrollDir, setScrollDir] = React.useState<GridMouseEventArgs["scrollEdge"]>();
 
+    // Adjust selection to include latest cell in visible region.
+    React.useEffect(() => {
```
I'm not sure this is right mechanism to do this. The easiest thing to do would be to find a way to reasonably re-emit the mouse move, triggering a new onItemHovered event as scroll is going on. The state update then would become a natural consequene of the artificial mousemove events.

BrianHung replied:
Makes sense. Most of the logic for adjusting the selection in this useEffect I just took from the `adjustSelection` function, but emitting artificial mousemove events is a better way of code re-usability and correctness.

BrianHung replied:
PR now triggers onItemHovered by listening to the `useAutoscroll` callback.

jassmith commented:
No pressure but I think this needs to change strategy if it can be merged. Let me know if you want me to keep this open or close it.

BrianHung commented:
Keep it open; looking into onItemHovered right now. Should it still be triggered using useEffect on scrollDir?

BrianHung commented:
Updated PR to use `onItemHoveredImpl` with `useEffect`. I wonder if it would even be easier if we modified `use-autoscroll` to accept a callback that's called every time the `scrollFn` is called. That would de-dupe the `useEffect` and `requestAnimationFrame`.

jassmith commented:
I mean I like less code, less to go wrong. Sounds like a good idea.

jassmith commented:
Can you rebase on 6.0.0. I had some bugs when you branched.

jassmith commented:
nvm turns out I could

jassmith commented:
I tested this, works great. Just need to grok how you made this work now.

jassmith commented on packages/core/src/data-editor/data-editor.tsx:
```diff
@@ -2447,8 +2446,11 @@ const DataEditorImpl: React.ForwardRefRenderFunction<DataEditorRef, DataEditorPr
+    const hoveredRef = React.useRef<GridMouseEventArgs>();
     const onItemHoveredImpl = React.useCallback(
         (args: GridMouseEventArgs) => {
+            if (mouseEventArgsAreEqual(args, hoveredRef.current)) return;
```
This is here to prevent spurious events being sent out to the user. Confirm?

BrianHung replied:
Yep; I put a console log statement there to check. It’s caused by the onItemHoveredImpl being called by onScroll.

BrianHung replied:
I basically copied the same logic from data-grid since it also has hoveredRef and onItemHovered.

jassmith reviewed (approved):
This looks amazing. Great work. We may need to do some consolidation work at some point with computing the out of bounds cell location but great work.
